README.md
# stata-exec

Send code to Stata from [Atom](https://atom.io). This package supports Windows, MacOS, and Linux.

---

While `stata-exec` will still be supported, I suggest users use the new [Stata Jupyter Kernel](https://kylebarron.github.io/stata_kernel/) along with the [Hydrogen package](https://atom.io/packages/Hydrogen) instead. It can show Stata results inline and works with Windows, macOS, and Linux.

![](./img/stata_kernel_example.gif)

---

![run-command](./img/run_command.gif)

## News

- Version 1.7.0 supports batch sending of code selection with `Run Batch`.
- Version 1.6.0 uses a new program on Linux to send code to Stata. It seems _way_ better; faster pasting and supports unicode characters.

## Installation

In the terminal run `apm install stata-exec` or go to Settings > Install and search for `stata-exec`.
This package depends on [`language-stata`](https://atom.io/packages/language-stata), which you should be prompted to install if needed.

### MacOS

MacOS has no dependencies but you must select the flavor of Stata you own in the package's configuration. Learn more in the [configuration section](#configuration) below. Then head to the [usage section](#usage) for more details on running code.

### Windows

Windows installation has a few steps, and I haven't been able to perfectly reproduce the working package on all computers. I hope to make this easier in the future, but for now this will have to do. Sadly, at this point **you need administrator privileges** to install this for Windows.

1. Make sure you've installed this package and `language-stata`. In the command prompt, run `apm install stata-exec language-stata` or go to Settings > Install and search for `stata-exec` and `language-stata`.
2. [Install this specific version of the program Node to your computer](https://nodejs.org/dist/v7.4.0/node-v7.4.0-x64.msi). Use the default installation settings.
3. Open up an administrator PowerShell (you can right click on the Windows icon at the bottom left and select "Windows PowerShell (Admin)") and type in:

    ```
    npm install --global --production windows-build-tools
    ```
    This took me 5-10 minutes to install. This is installing Python and other tools needed to install the package in the next step. When finished you should see a long list of names in a tree, like this:

    ```
    `-- windows-build-tools@2.2.1
      +-- chalk@2.3.2
      | +-- ansi-styles@3.2.1
      | | `-- color-convert@1.9.1
      | |   `-- color-name@1.1.3
    ...
    ```
4. Open up Command Prompt (type `cmd` in the search bar in the dock, and it will be the first result) and type in:

    ```
    cd %USERPROFILE%\.atom\packages\stata-exec
    npm install winax --python=%USERPROFILE%\.windows-build-tools\python27\python.exe
    atom -v
    ```

    Then enter the following, where you need to replace `ELECTRON_VERSION` with the text following "Electron" in the output of `atom -v`.

    ```
    npm rebuild winax --runtime=electron --target=ELECTRON_VERSION --disturl=https://atom.io/download/atom-shell --build-from-source
    ```

5. [Link the Stata Automation library](https://www.stata.com/automation/#install). The following steps worked for me on Windows 10. The Stata executable is most likely in the folder `C:\Program Files (x86)\Stata15`.

    > 1. In the installation directory, right-click on the Stata executable, for example, StataSE.exe. Choose "Create Shortcut".
    > 2. Right-click on the newly created "Shortcut to StataSE.exe", choose "Property", and change the Target from "C:\Program Files\Stata13\StataSE.exe" to "C:\Program Files\Stata13\StataSE.exe" /Register. Click "OK".
    > 3. Right-click on the updated "Shortcut to StataSE.exe"; choose "Run as administrator"

    While you're doing that, add the path of the Stata executable to the "Stata Path" option in the settings.

6. Restart Atom.

Now you can open up a Stata do-file and run code! See [Usage](#usage) for more details.

### Linux

**Important**: Linux dependencies changed in version 1.6.0. Linux users must install [xdotool](https://github.com/jordansissel/xdotool) and [xclip](https://github.com/astrand/xclip).

On Ubuntu:
```
sudo apt install xdotool xclip
```

## Usage

Code can be run using either the Command Palette or with keyboard shortcuts.

To open the Command Palette, press <kbd>cmd</kbd>-<kbd>shift</kbd>-<kbd>P</kbd> / <kbd>ctrl</kbd>-<kbd>shift</kbd>-<kbd>P</kbd>, and then start typing `Stata Exec`. The available commands will be shown in the drop-down menu.

The following are the default keyboard shortcuts (on macOS, use <kbd>cmd</kbd> instead of <kbd>ctrl</kbd>). Keyboard shortcuts can be personalized in your `keymap.cson` file. [More instructions here](http://flight-manual.atom.io/behind-atom/sections/keymaps-in-depth/).

| Command | Default keyboard binding                      | Description |
|-----------------------------|------------------|-----------------------------------------------------------------|
| `Run` | <kbd>ctrl</kbd>-<kbd>enter</kbd>              | Run selection or current line in Stata. |
| `Run All` | <kbd>shift</kbd>-<kbd>enter</kbd>             | Run selection or current line in Stata and move down. |
| `Run Batch` | <kbd>ctrl</kbd>-<kbd>alt</kbd>-<kbd>d</kbd> | Run selection with `do` instead of interactively pasting commands. |
| `Run And Move Down` | <kbd>shift</kbd>-<kbd>ctrl</kbd>-<kbd>D</kbd> | Run entire file in Stata. (This runs `do "/path/to/current/file"`) |
| `Run Paragraph` | <kbd>ctrl</kbd>-<kbd>alt</kbd>-<kbd>p</kbd>   | Run the previous command. |
| `Run Previous Command` | <kbd>shift</kbd>-<kbd>ctrl</kbd>-<kbd>C</kbd> | Change Stata's working directory to that of current file. |
| `Run Program` | <kbd>shift</kbd>-<kbd>ctrl</kbd>-<kbd>G</kbd> | Run paragraph under current cursor. A paragraph is a region enclosed by whitespace. |
| `Set Working Directory` | <kbd>shift</kbd>-<kbd>ctrl</kbd>-<kbd>R</kbd> | Run program definition under current cursor. See note below. |

For `Run Program`, if there exists `program drop` on the line before `program define`, the line including the former will be included in the selection. For example, all the lines in the below snippet would be sent to Stata:
```stata
cap program drop myProgram
program define myProgram
    // program contents
end
```

Note that `end` must be on a line by itself to be identified as the end of the program.

## Configuration

All configuration can be done in the settings panel (Settings > Packages > stata-exec). The available options are listed below:

- Stata Path (used for Windows only)
    - Absolute path to Stata executable. The default setting will most likely need to be changed to reflect your install location and Stata flavor.
- Which App (used for macOS only)
    - Select **StataIC**, **StataSE**, or **StataMP** depending on which version of Stata you have.
    - Select **XQuartz** if you want to run selections in session of Stata on a remote Unix server. To set this up, you need to have Stata already open in XQuartz; Atom will not open it for you. In your terminal, you'll need to do something like `ssh username@host -Y`, likely followed by `xstata`. This package's commands to run the entire do file and set the working directory are not supported on XQuartz.
- Advance Position
    - If checked, move cursor to the next line/paragraph after running the current line/paragraph.
- Focus Window
    - If checked, bring the Stata window to focus when sending code.
    - Otherwise, code runs in the background and the screen remains focused on Atom.
- Notifications
    - If checked, a pop-up notification will appear when a paragraph or function is not identified.
- Skip Comments
    - If this and Advance Position are checked, after running a line the cursor will move to the next uncommented line.
- Paste Speed (used for XQuartz only)
    - This value changes the amount of time the program waits between switching to the XQuartz window, pasting code, and sending "enter". The only way to send code to XQuartz is to use the clipboard, and the responsiveness of this process will depend on the speed of your internet connection. If the copy-pasting isn't working, try increasing the value. Decreasing the value will run your code faster. Value must be between 0.1 and 10.

## Notes

### Troubleshooting and Known Issues
- _Run All_ doesn't run the last line of the do file.
    - Stata needs there to be a _newline_ character following the last line of text. Add an empty line to the end of the file and it'll work.
- On Linux, the Stata GUI window must be the only program open with a window title of `Stata/`. If not, it appears the program will alternate between open Stata windows when pasting. xdotool seems to let me choose which Stata window to send text to, so maybe that could be put into a more useful function in the future.


CHANGELOG.md
# Changelog

## [1.8.6] - 2018-10-30

- Modify note about Stata Jupyter kernel in README.

## [1.8.5] - 2018-08-20

- Add note about Stata Jupyter kernel to README.

## [1.8.4] - 2018-08-20

- Add `Stata` as option in Mac executable list.

## [1.8.3] - 2018-08-07

- Remove comments when sending paragraphs and programs.

## [1.8.2] - 2018-08-02

- Fix running on Windows

## [1.8.1] - 2018-07-30

- Fix batch end of line errors. #41
- Fix menus

## [1.8.0] - 2018-07-30

- Support multiple cursors. #34
- Fix batch sending of code. #41
- Rewrite much of package in better JavaScript.

## [1.7.0] - 2018-07-25

- Support batch sending of code selection; run with `do`.
- On Linux, print stdout and stderr of xdotool to console
- On Linux, fix bug where xdotool (more likely xclip) would fail when the clipboard was empty, i.e. upon first logging in.

## [1.6.0] - 2018-07-24

- Use xdotool instead of autokey on Linux
    - Improves paste speed, reliability; allows for pasting unicode characters
    - Allows you to type while the code is pasting
    - In future could create functionality to support two Stata windows at once
- Important bugfixes for using xdotool
    - Search only for the first Stata window
    - Force <kbd>ctrl</kbd> to be not pressed when the program sends <kbd>ctrl</kbd>+<kbd>v</kbd>. Since the user usually is pressing <kbd>ctrl</kbd> to run the command, this is important. Otherwise all that is entered into Stata is <kbd>v</kbd>.


## [1.5.2] - 2018-06-01

- Update keymaps so the package works within Dynamic Documents

## [1.5.1] - 2018-05-17
- Require that `end` be on a line by itself to indicate the end of a program.
    This is to prevent false positives where the word "end" appears elsewhere.

## [1.5.0] - 2018-05-17
- Update API to resemble Hydrogen's API (keeping current keyboard shortcuts)
- Add `Run and Move Down` command, bound to <kbd>shift</kbd>-<kbd>enter</kbd>
- Fix bug in Linux where sending a large selection pasted some text in Atom instead of Stata.

## [1.4.6] - 2018-03-23
- Fix bug that prevented graph windows from opening

## [1.4.5] - 2018-03-22
- Add final `\r` to code on Windows

## [1.4.4] - 2018-03-22
- Remove `\r\n` when removing comments, not just `\n`, to support windows line endings

## [1.4.3] - 2018-03-09
- On Windows, when opening a new Stata session, set Stata's working directory to current file's working directory

## [1.4.2] - 2018-03-09
- Small additions to README

## [1.4.1] - 2018-03-09
- Try loading winax in a try/catch clause

## [1.4.0] - 2018-03-09
- Add Windows support!

## [1.3.2] - 2018-02-23
- Add try-except clause in `linux/stata-exec.py` to hopefully work correctly on Arch Linux.

## [1.3.1] - 2018-01-31
- Same as 1.3.0 but didn't publish correctly the first time

## [1.3.0] - 2018-01-31
- Fix running of multiple lines with ///

## [1.2.1] - 2017-11-20
- Fix bug that couldn't find removeComments

## [1.2.0] - 2017-11-20
- Add linux support.
- Remove comments for all methods of sending code, not just 'send-command'
- Change warnings from "can't find function" to "can't find program"

## [1.1.2] - 2017-11-19
- Fix comment removal to not touch comments in strings. Fixes https://github.com/kylebarron/stata-exec/issues/2
    - Thanks to https://stackoverflow.com/questions/24518020/comprehensive-regexp-to-remove-javascript-comments

## [1.1.1] - 2017-11-19
- Add restriction that code sent through Applescript can be max 8192 characters. https://www.stata.com/automation/#list
- Include a line with `program drop myProgram` if it comes before `program define myProgram` when sending a program to Stata.

## [1.1.0] - 2017-11-19
- Add XQuartz support. This allows code to be run on macOS in a session of Stata running on a remote server.

## [1.0.3] - 2017-11-15
- Fix the "Set Working Directory" command to use compound double quotes.
- Refactor "Do Entire File" to use `addSlashes()`
- Fix links in README.md; some other updates.

## [1.0.1] and [1.0.2] - 2017-11-13
- No updates; issues getting atom.io to publish the package.

## [1.0.0] - 2017-11-13
- Atom [stata-exec](https://atom.io/packages/stata-exec) package transferred to [@kylebarron](https://github.com/kylebarron) and hosted at https://github.com/kylebarron/stata-exec.
- Ported from version 0.5.0 of [atom-r-exec](https://github.com/pimentel/atom-r-exec). All functionality ported:
    - cmd-enter: send selection or current line to Stata.
    - shift-cmd-d: send entire file to Stata. (File must be saved first. This runs do "/path/to/current/file")
    - shift-alt-p: send the previous command.
    - shift-cmd-c: change Stata's working directory to that of current file.
    - shift-cmd-g: send paragraph under current cursor. A paragraph is a region enclosed by whitespace.
    - shift-cmd-r: send program definition under current cursor. For example, all the lines in the below snippet would be sent to Stata:

## [0.2.0] - 2017-06-27
- Original code created by [@nickeubank](https://github.com/nickeubank). Still in repository located at https://github.com/nickeubank/atom-stata-exec.


lib/sendCode.js
'use strict';

const { spawn } = require('child_process');
const delay = require('delay');
const path = require('path');
const fs = require('fs');
const runApplescript = require('run-applescript');
const escapeString = require('escape-string-applescript');

module.exports = {
  previousCommand: '',
  winax: null,

  send(text, batch=false) {
    console.log('entering sendCode function');
    this.previousCommand = text;
    switch (process.platform) {
      case 'darwin':
        return this.sendMac(text);
      case 'linux':
        return this.sendLinux(text);
      case 'win32':
        return this.sendWindows(text);
      default:
        return this.error('Unknown operating system.');
    }
  },

  error(message) {
    console.error(message);
    return atom.notifications.addError(message);
  },

  defineWinax(){
    console.log('requiring winax')
    const os = require('os');
    var winaxPath = os.homedir();
    winaxPath += "/.atom/packages/stata-exec/node_modules/winax";
    try {
      return require(winaxPath);
    } catch (err) {
      console.error(err);
      atom.notifications.addError('Winax not correctly installed');
    }
  },

  sendWindows(text) {
    if (text.length > 8192) {
      this.error('Code to send must be <= 8192 characters');
      return;
    }

    if (!this.winax) {
      this.winax = this.defineWinax();
    }

    if (this.con) {
      return this.con.DoCommandAsync(text);
    }

    spawn(atom.config.get('stata-exec.stataPath'), { stdio: 'ignore', detached: true }).unref();
    return delay(2000)
      .then(() => {
        try {
          this.con = new this.winax.Object("stata.StataOLEApp", { // jshint ignore:line
            activate: true,
            async: true,
            type: true
          });
        } catch(err) {
          console.error(err);
          if (err.message.includes('Invalid class string')) {
            return this.error('Registering stata failed');
          }
        }
        const editor = atom.workspace.getActiveTextEditor();
        var cwd = editor.getPath();
        var folderPath = cwd.substring(0, cwd.lastIndexOf('\\') + 1);
        this.con.DoCommandAsync('cd `"' + folderPath + '"\'');
        this.con.DoCommandAsync(text);
      });
  },

  sendMac(text) {
    text = escapeString(text);
    const whichApp = atom.config.get('stata-exec.whichApp');
    if (whichApp === 'XQuartz') {
      return this.sendXQuartz(text);
    }

    if (text.length > 8192) {
      this.error('Code to send must be <= 8192 characters');
      return;
    }

    const focusWindow = atom.config.get('stata-exec.focusWindow');
    var cmd = '';
    if (focusWindow) {
      cmd += `tell application "${whichApp}" to activate\n`;
    }
    cmd += `tell application "${whichApp}" to DoCommandAsync "${text}"`;

    runApplescript(cmd)
      .then(() => {
        console.log('Finished applescript');
        console.log('code: ', text);
        return console.log('Applescript: ', cmd);
      }).catch((err) => {
        console.error(err);
        console.error('code: ', text);
        return console.error('Applescript: ', cmd);
      });
  },

  sendLinux(text) {
    var exec = require('child_process').exec;
    // The `keyup ctrl` is the most important part of this
    //
    // The --clearmodifiers flag doesn't work great. By default, I have most of
    // the run keys bound with a `ctrl` key. I.e. to run a line of code you'd
    // type `ctrl+enter`. But when `ctrl` is held down by the user, running
    // `ctrl+v` pastes only `v`, and doesn't paste the clipboard. To prevent
    // against this, I use `keyup ctrl`, which forces the beginning position of
    // `ctrl` to be not pressed, regardless of what the user is doing.
    var cmd = `
      old_cb="$(xclip -o -selection clipboard)";
      this_window="$(xdotool getactivewindow)" &&
      stata_window="$(xdotool search --name --limit 1 "Stata/(IC|SE|MP)? 1[0-9]\.[0-9]")" &&
      cat ~/.stata-exec_code | xclip -i -selection clipboard &&
      xdotool \
        keyup ctrl shift \
        windowactivate --sync $stata_window \
        key --clearmodifiers --delay 100 ctrl+v Return \
        windowactivate --sync $this_window;
      printf "$old_cb" | xclip -i -selection clipboard`;

    var codepath = path.join(process.env.HOME, '.stata-exec_code');
    fs.writeFile(codepath, text, function(err) {
      if(err) {
        console.log(err);
        this.error('Home directory not writeable. Check permissions');
        return;
      }
      console.log('The file was saved!');
      exec(cmd, function(err, stdout, stderr) {
        console.log('stdout: ' + stdout);
        console.log('stderr: ' + stderr);
        if (err) throw err;
      });
    });
  },

  sendXQuartz(text) {
    const pasteSpeed = atom.config.get('stata-exec.pasteSpeed');
    const focusWindow = atom.config.get('stata-exec.focusWindow');
    var cmd = `
      set current_clipboard to the clipboard
      set the clipboard to ("${text}" as text)
      tell application "XQuartz" to activate
      delay 0.4 * ${pasteSpeed}
      tell application "System Events" to keystroke "v" using control down
      delay 0.9 * ${pasteSpeed}
      tell application "System Events" to keystroke return
      delay 0.1 * ${pasteSpeed}`;
    if (!focusWindow) {
      cmd += '\ntell application "Atom" to activate';
    }
    cmd += '\nset the clipboard to current_clipboard';

    runApplescript(cmd)
      .then(() => {
        console.log('Finished applescript');
      }).catch((err) => {
        console.error(err);
        console.error('code: ', text);
        return console.error('Applescript: ', cmd);
      });
  }
};


linux/stata-exec.py
# Enter script code
import time
from os.path import expanduser, join
shortdelay = 0.2
current = window.get_active_title()

with open(join(expanduser('~'), '.stata-exec_code'), 'r') as f:
    cmd = f.read()

block_size = 100
cmd_blocks = [cmd[i:i + block_size] for i in range(0, len(cmd), block_size)]

window.activate("Stata/")

try:
    if window.wait_for_focus("Stata/(IC|SE|MP) 1[1-5].[0-2]", timeOut = 1):

        for block in cmd_blocks:
            time.sleep(shortdelay)
            keyboard.send_keys(block)

        time.sleep(shortdelay)
        keyboard.send_keys("<enter>")

        time.sleep(shortdelay)
        window.activate(current)
except TypeError:

    for block in cmd_blocks:
        time.sleep(shortdelay)
        keyboard.send_keys(block)

    time.sleep(shortdelay)
    keyboard.send_keys("<enter>")

    time.sleep(shortdelay)
    window.activate(current)


lib/main.js
'use strict';

const { CompositeDisposable, Point, Range } = require('atom');
const path = require('path');
const fs = require('fs');
const sendCode = require('./sendCode');

const apps = {
  stataMP: 'StataMP',
  stata: 'Stata',
  stataIC: 'StataIC',
  stataSE: 'StataSE',
  xquartz: 'XQuartz'
};

module.exports = {
  config: {
    whichApp: {
      type: 'string',
      enum: [apps.stataMP, apps.stataSE, apps.stata, apps.stataIC, apps.xquartz],
      default: apps.stataSE,
      description: 'Which application to send code to. Only applicable for MacOS. The StataIC executable is currently named `Stata`, so try choosing that if you\'re having issues.',
      order: 2
    },
    advancePosition: {
      type: 'boolean',
      default: false,
      order: 4,
      description: 'Cursor advances to the next line after ' +
        'sending the current line when there is no selection'
    },
    skipComments: {
      type: 'boolean',
      default: true,
      order: 5,
      description: 'When "advancePosition" is true, skip lines that contain ' +
        'only comments'
    },
    focusWindow: {
      type: 'boolean',
      default: true,
      order: 3,
      description: 'After code is sent, bring focus to where it was sent'
    },
    pasteSpeed: {
      type: 'number',
      default: 1.0,
      minimum: 0.1,
      maximum: 10,
      order: 7,
      description: 'This is only applicable for XQuartz. This value changes the amount of time the program waits between switching to the XQuartz window, pasting code, and sending "enter". The only way to send code to XQuartz is to use the clipboard, and the responsiveness of sending code will depend on the speed of your internet connection. If the copy-pasting isn\'t working, try increasing the value. Decreasing the value will run your code faster. Value must be between 0.1 and 10.'
    },
    stataPath: {
      type: 'string',
      description: 'Absolute path to Stata executable. Only applicable for Windows.',
      default: 'C:\\Program Files (x86)\\Stata15\\StataSE-64.exe',
      order: 1,
    },
    allowSave: {
      type: 'boolean',
      default: true,
      description: 'Save file automatically before running "Run All"'
    }
  },

  subscriptions: null,

  activate(state) {
    this.subscriptions = new CompositeDisposable();

    this.subscriptions.add(atom.commands.add('atom-text-editor',
      'stata-exec:run', () => this.run()));
    this.subscriptions.add(atom.commands.add('atom-text-editor',
      'stata-exec:run-and-move-down', () => this.run(true)));
    this.subscriptions.add(atom.commands.add('atom-text-editor',
      'stata-exec:run-previous-command', () => this.runPreviousCommand()));
    this.subscriptions.add(atom.commands.add('atom-text-editor',
      'stata-exec:run-all', () => this.runAll()));
    this.subscriptions.add(atom.commands.add('atom-text-editor',
      'stata-exec:run-batch', () => this.run(false, true)));
    this.subscriptions.add(atom.commands.add('atom-text-editor', 'stata-exec:run-paragraph', () => this.runParagraph()));
    this.subscriptions.add(atom.commands.add('atom-text-editor', 'stata-exec:run-program', () => this.runFunction()));
    return this.subscriptions.add(atom.commands.add('atom-text-editor',
      'stata-exec:set-working-directory', () => this.setWorkingDirectory()));
  },

  deactivate() {
    return this.subscriptions.dispose();
  },

  _getEditorAndBuffer() {
    const editor = atom.workspace.getActiveTextEditor();
    const buffer = editor.getBuffer();
    return [editor, buffer];
  },

  runAll() {
    const whichApp = atom.config.get('stata-exec.whichApp');
    var msg = '';
    if (whichApp == "XQuartz") {
      this.error('Running entire do file not supported for XQuartz');
      return;
    }

    const [editor, buffer] = this._getEditorAndBuffer();
    if (atom.config.get('stata-exec.allowSave')) {
      editor.save();
    }
    const documentTitle = editor.getPath();
    if (!documentTitle) {
      this.error('Error: File not yet saved.');
      return;
    }
    const doFileCommand = `do \`"${documentTitle}"'`;
    return sendCode.send(doFileCommand);
  },

  run(advancePosition=false, batch=false) {
    const whichApp = atom.config.get('stata-exec.whichApp');
    const [editor, buffer] = this._getEditorAndBuffer();
    // we store the current position so that we can jump back to it later
    // (if the user wants to)
    const currentPosition = editor.getLastSelection().getScreenRange().end;
    const selections = this.getSelections();
    const texts = selections.texts;

    for (var i = 0; i < texts.length; i++) {
      if (batch) {
        this.runBatch(texts[i]);
      } else {
        // For interactive code (i.e. copy-pasted code) need to make sure there
        // are no block comments or ///.
        sendCode.send(this.removeComments(texts[i]));
      }
    }

    if (atom.config.get('stata-exec.advancePosition') === true) {
      advancePosition = true;
    }
    if (advancePosition && !selections.anySelection) {
      let nextPosition = this._findForward(this.nonEmptyLine, currentPosition.row + 1);
      if (nextPosition != null) {
        if (nextPosition == null) { nextPosition = [currentPosition + 1, 0]; }
        editor.setCursorScreenPosition(nextPosition);
        return editor.moveToFirstCharacterOfLine();
      }
    } else {
      if (!selections.anySelection) {
        return editor.setCursorScreenPosition(currentPosition);
      }
    }
  },

  runBatch(text) {
    var codepath = path.join(process.env.HOME, '.stata-exec_batch_code');
    if (process.platform == 'win32') {
      var eol = '\r\n'
    } else {
      var eol = '\n'
    }
    if (text.substr(text.length - 1) !== '\n') {
      text += eol;
    }
    fs.writeFile(codepath, text, function(err) {
      if(err) {
        return console.log(err);
      }
      console.log(`The batch code was saved to ${codepath}!`);
      sendCode.send(`do \`"${codepath}"'`);
    });
  },

  removeComments(text) {
    console.log('code with comments: ' + text);
    text = text.replace(/((["'])(?:\\[\s\S]|.)*?\2|(?:[^\w\s]|^)\s*\/(?![*\/])(?:\\.|\[(?:\\.|.)\]|.)*?\/(?=[gmiy]{0,4}\s*(?![*\/])(?:\W|$)))|\/\/\/.*?\r?\n\s*|\/\/.*?$|\/\*[\s\S]*?\*\//gm, '$1');
    // https://stackoverflow.com/questions/24518020/comprehensive-regexp-to-remove-javascript-comments
    // Using the "Final Boss Fight" at the bottom. Otherwise it fails on `di 5 / 5 // hello`
    // code = code.replace(';', '')
    if (process.platform == 'win32') {
      text = text + '\r';
    }
    console.log('code without comments: ' + text);
    return text;
  },

  runPreviousCommand() {
    const whichApp = atom.config.get('stata-exec.whichApp');
    return sendCode.send(sendCode.previousCommand);
  },

  getFunctionRange() {
    // gets the range of the closest function above the cursor.
    // if there is no (proper) function, return false
    const [editor, buffer] = this._getEditorAndBuffer();
    const currentPosition = editor.getCursorBufferPosition();
    // search for the simple function that looks something like:
    // label <- function(...) {
    // in case the current function definition is on the current line
    currentPosition.row += 1;
    const backwardRange = [0, currentPosition];
    const funRegex = new RegExp(/^\s*(pr(ogram|ogra|ogr|og|o)?)\s*(?!drop\s+)(de(fine|fin|fi|f)?)?\s*[A-Za-z_][A-Za-z0-9_]{0,31}/g);
    let foundStart = null;
    editor.backwardsScanInBufferRange(funRegex, backwardRange, function(result) {
      if (result.range.start.column === 0) {
        foundStart = result.range;
        return result.stop();
      }
    });

    if ((foundStart == null)) {
      console.error("Couldn't find the beginning of the program.");
      return null;
    }

    const dropRegex = new RegExp(/\s*pr(ogram|ogra|ogr|og|o)?\s+(drop)\s+[A-Za-z_][A-Za-z0-9_]{0,31}/g);
    const textPrevRow = editor.lineTextForBufferRow(foundStart.start.row - 1);
    if (dropRegex.test(textPrevRow) == true) {
      foundStart.start.row -= 1;
    }

    // now look for the end
    const numberOfLines = editor.getLineCount();
    const forwardRange = [foundStart.start, new Point(numberOfLines + 1, 0)];

    let foundEnd = null;
    editor.scanInBufferRange(/^\s*end\s*$/g, forwardRange, function(result) {
      if (result.range.start.column === 0) {
        foundEnd = result.range;
        return result.stop();
      }
    });

    if ((foundEnd == null)) {
      console.error("Couldn't find the end of the program.");
      return null;
    }

    // check if cursor is contained in range
    currentPosition.row -= 1;
    if ((foundStart.start.row <= currentPosition.row) &&
      (currentPosition.row <= foundEnd.start.row)) {
      return new Range(foundStart.start, foundEnd.end);
    } else {
      console.error("Couldn't find a program surrounding the current line.");
      console.error("start: ", foundStart);
      console.error("end: ", foundEnd);
      console.error("currentPosition: ", currentPosition);
      return null;
    }
  },

  runFunction() {
    const [editor, buffer] = this._getEditorAndBuffer();
    const whichApp = atom.config.get('stata-exec.whichApp');

    const range = this.getFunctionRange();
    if (range != null) {
      let code = editor.getTextInBufferRange(range);
      return sendCode.send(this.removeComments(code));
    } else {
      return this.Warning("Couldn't find program.");
    }
  },

  getSelections() {
    // returns an object with keys:
    // selection: the selection or line at which the cursor is present
    // anySelection: if true, the user made a selection.
    const [editor, buffer] = this._getEditorAndBuffer();

    let selections = editor.getSelectionsOrderedByBufferPosition();
    let texts = Array(selections.length);
    let anySelection = true;

    for (var i = 0; i < selections.length; i++) {
      var selection = selections[i];
      if (selection.getText() === '') {
        anySelection = false;
        const cursorPosition = selection.cursor.getBufferPosition().row;
        texts[i] = editor.lineTextForBufferRow(cursorPosition);
      } else {
        texts[i] = selection.getText();
      }
    }

    return { texts, anySelection };
  },

  error(message) {
    console.error(message);
    return atom.notifications.addError(message);
  },

  warning(message) {
    console.error(message);
    return atom.notifications.addWarning(message);
  },

  onlyWhitespace(str) {
    // returns true if string is only whitespace
    return str.replace(/\s/g, '').length === 0;
  },

  getCurrentParagraphRange() {
    let lineIndex;
    const [editor, buffer] = this._getEditorAndBuffer();
    const currentPosition = editor.getCursorBufferPosition().row;

    let currentLine = buffer.lineForRow(currentPosition);

    if (this.onlyWhitespace(currentLine)) {
      return null;
    }

    let startIndex = -1;
    // if we exhaust loop, then this paragraph begins at the first line
    if (currentPosition > 0) {
      let asc, start;
      for (start = currentPosition - 1, lineIndex = start, asc = start <= 0; asc ? lineIndex <= 0 : lineIndex >= 0; asc ? lineIndex++ : lineIndex--) {
        currentLine = buffer.lineForRow(lineIndex);
        if (this.onlyWhitespace(currentLine)) {
          startIndex = lineIndex;
          break;
        }
      }
    }
    startIndex += 1;

    let endIndex = editor.getLineCount();
    const numberOfLines = editor.getLineCount() - 1;
    if (currentPosition < (endIndex - 1)) {
      let asc1, end, start1;
      for (start1 = currentPosition + 1, lineIndex = start1, end = numberOfLines, asc1 = start1 <= end; asc1 ? lineIndex <= end : lineIndex >= end; asc1 ? lineIndex++ : lineIndex--) {
        currentLine = buffer.lineForRow(lineIndex);
        if (this.onlyWhitespace(currentLine)) {
          endIndex = lineIndex;
          break;
        }
      }
    }
    endIndex -= 1;

    const paragraphRange = new Range([startIndex, 0], [endIndex, buffer.lineLengthForRow(endIndex)]);

    return paragraphRange;
  },

  runParagraph() {
    const whichApp = atom.config.get('stata-exec.whichApp');
    const [editor, buffer] = this._getEditorAndBuffer();
    const paragraphRange = this.getCurrentParagraphRange();

    if (paragraphRange) {
      let code = editor.getTextInBufferRange(paragraphRange);
      sendCode.send(this.removeComments(code));
      const advancePosition = atom.config.get('stata-exec.advancePosition');
      if (advancePosition) {
        const currentPosition = editor.getLastSelection().getScreenRange().end;
        let nextPosition = this._findForward(this.nonEmptyLine, paragraphRange.end.row + 1);
        if (nextPosition != null) {
          if (nextPosition == null) { nextPosition = [currentPosition + 1, 0]; }
          editor.setCursorScreenPosition(nextPosition);
          return editor.moveToFirstCharacterOfLine();
        }
      }
    } else {
      return this.Warning('No paragraph at cursor.');
    }
  },

  nonEmptyLine(line) {
    const skipComments = atom.config.get('stata-exec.skipComments');
    let ret = true;
    if (skipComments) {
      ret = !/^\s*#/.test(line);
    }
    // a non empty line is a line that doesn't contain only a comment
    // and at least 1 character
    return ret && /\S/.test(line);
  },

  _findForward(searchFun, startPosition = null) {
    const editor = atom.workspace.getActiveTextEditor();
    const buffer = editor.getBuffer();

    if ((startPosition == null)) {
      startPosition = editor.getCursorBufferPosition().row;
    }

    let index = null;
    const numberOfLines = editor.getLineCount() - 1;
    if (startPosition >= numberOfLines) {
      return null;
    }
    for (let lineIndex = startPosition, end = numberOfLines, asc = startPosition <= end; asc ? lineIndex <= end : lineIndex >= end; asc ? lineIndex++ : lineIndex--) {
      const currentLine = buffer.lineForRow(lineIndex);
      if (searchFun(currentLine)) {
        index = lineIndex;
        break;
      }
    }

    if (index != null) {
      return [index, buffer.lineLengthForRow(index)];
    }

    return null;
  },

  setWorkingDirectory() {
    const whichApp = atom.config.get('stata-exec.whichApp');
    if (whichApp == "XQuartz") {
      this.Warning('Set Working Directory not supported for XQuartz');
      return;
    }

    // set the current working directory to the directory of
    // where the current file is
    const [editor, buffer] = this._getEditorAndBuffer();
    let cwd = editor.getPath();
    if (!cwd) {
      this.Warning('No current working directory (save the file first).');
      return;
    }
    cwd = cwd.substring(0, cwd.lastIndexOf('/'));
    cwd = `cd \`"${cwd}"'`;

    return sendCode.send(cwd);
  }
};
